### ISSUE TITLE
Operation specific dependency injection

### ISSUE THREAD

**Issue by servernoj** [danielgtaylor/huma#293]
Hello,

I am working on integration testing, where every operation is tested against a dedicated DB instance running in docker (with the help of `ory/dockertest` and `testify/suite`). Due to the non-parallel nature of `testify/suite`, I can run tests in sequence and override the global DB handle before every test by analogy to https://github.com/danielgtaylor/huma/issues/198 -- that works

I want to run those tests **in parallel**. I created a simplified version of `testify/suite` in which individual tests (from the same suite) do not rely on shared data but instead use storage keyed to the test name. It allows multiple DB containers to co-exist. 

I now need to figure out a way to feed individual operations with specific DB handles, such that every test in the suit could pass (inject) its associated DB handle into a specific testable operation. In this case, when multiple tests run in parallel, each with its own DB running in a container, the associated operations would read/write data from the correct DB. 

So far, I think in terms of extending #198 and associating multiple DB handles to the "service" struct (where individual operations are registered) such that every operation could "pick" a correct handle (while being tested) or all would pick some default one (while in production). The setup for "picking" becomes overcomplicated because operations need to distinguish testing from production. 

**My main question is**: Is what I'm doing at all possible?

PS: My desire to run tests in parallel is based on curiosity only; I found that `testify/suite` does not (if I'm not mistaken) support parallel testing, and I just wanted to challenge myself...

Thank you

**Comment by danielgtaylor** [comment]
So if we take some simplified code like from #198:

```go
type DB interface {
	Get(query string) any
	Put(query string)
}

type DemoServer struct {
	db DB
}

func NewDemoServer(db DB) *DemoServer {
	return &DemoServer{
		db: db,
	}
}

func (s *DemoServer) RegisterDemo(api huma.API) {
	huma.Get(api, "/demo", func(ctx context.Context, input *struct{}) (*struct{}, error) {
		// TODO: use dependencies here as s.db
		return nil, nil
	})
}

func main() {
	db := NewDB()
	demo := NewDemoServer(db)
	
	router := chi.NewMux()
	api := humachi.New(router, huma.DefaultConfig("My API", "1.0.0"))

	huma.AutoRegister(api, demo)

	http.ListenAndServe(":8888", router)
}
```

Then using something like [mockery](https://vektra.github.io/mockery/latest/) you can add a generate command for the DB interface:

```go
//go:generate mockery --name DB
```

Then run `go generate` and then write your tests however you want:

```go
func TestDemoServer(t *testing.T) {
	db := NewMockDB()
	db.EXPECT().Get("some query").Return("some value")

	demo := NewDemoServer(db)
	router, api := humatest.New(t)
	huma.AutoRegister(api, demo)

	resp := api.Get("/demo")
	assert.Equal(t, http.StatusNoContent, resp.StatusCode)
}
```

Since these tests just use a mock of the DB they can all run in parallel without any issues. Hope that helps!

**Comment by servernoj** [comment]
@danielgtaylor, thank you for the hint, but this is slightly different than what I needed. In my vision, it will be one instance of `DemoServer` (using terminology from the example) with multiple operations -- each with its own DB. 

But in the solution that you propose (if I see it right), we create multiple instances of `DemoServer,` each with a single DB and a single Huma operation. 

Maybe this is a way to go... as I mentioned earlier, my point was to see the feasibility of a particular setup:
- one instance of `DemoServer`
- multiple `auto-registrated` operation
- using shared DB when `DemoServer` is created for production
- using 1:1 of DB to operation binding when it is used for integration testing
- for the purpose of that integration testing, one operation is scoped for a single test (with multiple sub-tests) and is associated with one DB

**Comment by servernoj** [comment]
I think I solved the challenge by making `DemoServer` implement some `Deps` (dependencies) interface, which has methods to `set` and `get` arbitrary dependencies in the context of a specific operation. The concrete implementation of that interface has storage protected by `sync.RWMutex`. 

For prod, I instantiate `Deps` into `DemoServer` with some defaults, e.g., with a shared DB handle. But for testing, I set it to store the test-specific version of DB (and other dependencies).
